crates/ty_python_semantic/src/node_key.rs

```diff
@@ -1,21 +1,14 @@
-use ruff_python_ast::AnyNodeRef;
+use ruff_python_ast::HasNodeIndex;
 
 /// Compact key for a node for use in a hash map.
-///
-/// Stores the memory address of the node, because using the range and the kind
-/// of the node is not enough to uniquely identify them in ASTs resulting from
-/// invalid syntax. For example, parsing the input `for` results in a `StmtFor`
-/// AST node where both the `target` and the `iter` field are `ExprName` nodes
-/// with the same (empty) range `3..3`.
 #[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
 pub(super) struct NodeKey(usize);
```

Comment:
Should we just store the `NodeIndex` here. That would also ensure that the type now shrinks to a `u32`. It probably even raises the question if `NodeKey` is still needed. It would probably be a good exercise to go through all uses and see where it can be replaced (maybe even with a `Vec` that directly indexes into the Nodes array)

Reply:
Yeah I will do this once indices are assigned by the parser. Currently `NodeIndex` is `!Copy`.